feat: add onShiftClickOnly to multiSort

niiyeboah opened the pull request:
## Description

Please list all relevant dependencies in this section and provide summary of the change, motivation and context.

Fixes vaadin/platform#3378

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/flow-components?pullRequest=4005) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/flow-components?pullRequest=4005) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/flow-components?pullRequest=4005) it.</sub>

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewSortingIT.java:
```diff
@@ -184,6 +189,26 @@ public void gridWithSorting_multiSortPriorityAppend() {
                 QuerySortOrder.asc("firstName").thenDesc("age").build(), true);
     }
 
+    @Test
```
nit: since the change just adds a bit of server-side logic, I assume it could be tested with a couple of unit tests. IT feels like overkill for the purpose if there's no flow-components client-side logic involved.

niiyeboah replied:
Ah ok got it. I've removed the ITs and added unit tests now

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -3209,13 +3209,49 @@ public void setMultiSort(boolean multiSort) {
+     *            {@code false} for normal multi-sort behavior
+     */
+    public void setMultiSort(boolean multiSort, boolean onShiftClickOnly) {
+        getElement().setAttribute("multi-sort-on-shift-click",
+                multiSort && onShiftClickOnly);
```
Is the check for multiSort necessary here? Could it just be `getElement().setAttribute("multi-sort-on-shift-click", onShiftClickOnly);`?

It would be preferable to operate directly with the properties instead of attributes (which get parsed as properties by Polymer): `getElement().setProperty("multiSortOnShiftClick", onShiftClickOnly);`

niiyeboah replied:
> Is the check for multiSort necessary here? Could it just be getElement().setAttribute("multi-sort-on-shift-click", onShiftClickOnly);?
I am not sure about this to be honest. My thinking is that it would cause confusion if the user is able to `setMultiSort(false, true)`, thinking they changed multi-sort to work on shift-click only but they have disabled multi-sort. If that makes sense. Actually I will add extra documentation to clarify this and it should solve any problems.

tomivirkki replied:
Fair enough. My reasoning for this was that on the Web Component: `<vaadin-grid multi-sort-on-shift-click>` also works even without adding `multi-sort`

niiyeboah replied:
Removed the extra check, added more documentation and refactored to use property instead of attribute.

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -3236,6 +3272,23 @@ public boolean isMultiSort() {
+     *         {@code false} otherwise
+     */
+    public boolean isMultiSortShiftClickOnly() {
+        String onShiftClickOnly = getElement()
+                .getAttribute("multi-sort-on-shift-click");
+        if (onShiftClickOnly != null && onShiftClickOnly.length() == 0) {
+            onShiftClickOnly = "true";
+        }
+        return isMultiSort() && Boolean.parseBoolean(onShiftClickOnly);
```
Would it be enough to `return getElement().hasAttribute("multi-sort-on-shift-click");` (or rather `return getElement().getProperty("multiSortOnShiftClick", false);`)?

niiyeboah replied:
Yes and done.

niiyeboah replied:
One thing to note is that I was copying the `isMultiSort` implementation so maybe that needs to be changed in the future for consistency.

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridSortingTest.java:
```diff
@@ -293,6 +293,27 @@ public void setMultiSortWithPriorityNull_throws() {
+    @Test
+    public void setMultiSortShiftClickOnly() {
+        grid.setMultiSort(true, true);
+
```
Please also check `grid.getElement().hasAttribute("multi-sort-on-shift-click");` (or rather the `multiSortOnShiftClick` property).

An additional test could check the case when the attribute is not there and `isMultiSortShiftClickOnly` returns false.

niiyeboah replied:
Done.

niiyeboah commented:
I rebased the branch to get the latest changes from master and have also changed `setMultiSort` to require `multiSort` to be true in order for `onShiftClickOnly` to be enabled.

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -3205,13 +3205,51 @@ public void setMultiSort(boolean multiSort) {
+     *            behavior
+     */
+    public void setMultiSort(boolean multiSort, boolean onShiftClickOnly) {
+        getElement().setProperty("multiSortOnShiftClick",
```
The following case still has issues:
- Set multi-sort true with onShiftClickOnly
- Set multi-sort false (without using onShiftClickOnly APIs)

The result is that the Grid is still multi-sortable with shift-click:

```java
grid.setMultiSort(true, true);
grid.setMultiSort(false);

Assert.assertEquals("false", grid.getElement().getProperty("multiSortOnShiftClick"));
```

niiyeboah replied:
Changed to account for this case.

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridSortingTest.java:
```diff
@@ -293,6 +293,40 @@ public void setMultiSortWithPriorityNull_throws() {
+    public void setMultiSortShiftClickOnly() {
+        grid.setMultiSort(true, true);
+
+        Assert.assertEquals("true",
+                grid.getElement().getProperty("multiSortOnShiftClick"));
```
Prefer using `assertTrue` and `assertFalse` in the tests:

For example
```java
Assert.assertTrue(grid.getElement().getProperty("multiSortOnShiftClick", false));
```

niiyeboah replied:
Done.

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -3205,13 +3208,56 @@ public void setMultiSort(boolean multiSort) {
+        updateMultiSortPriority(priority);
+    }
+
+    private void updateMultiSortShiftClickOnly(Boolean onShiftClickOnly) {
```
Use primitive `boolean` instead of boxed `Boolean`

niiyeboah replied:
Done.

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -3205,13 +3208,56 @@ public void setMultiSort(boolean multiSort) {
+        updateMultiSortPriority(priority);
+    }
+
+    private void updateMultiSortShiftClickOnly(Boolean onShiftClickOnly) {
```
Would rename this as `updateMultiSortOnShiftClick` (also the param name) since that's effectively the property it updates

niiyeboah replied:
Done.

tomivirkki replied:
No, I meant only the private method that just sets the property. The public API does more than that so it should not be renamed.

tomivirkki replied:
Looking at the diff, I noticed it's missing the `On` word in the middle. So it could be `grid.isMultiSortOnShiftClickOnly`.

tomivirkki replied:
Or should we drop the getter? I didn't see it mentioned in the API design at https://github.com/vaadin/platform/issues/3378
